src/sage/matrix/matrix_polynomial_dense.pyx

```diff
@@ -4079,3 +4084,564 @@ cdef class Matrix_polynomial_dense(Matrix_generic_dense):
+                     [5*x^2 + 4*x,   3*x^2 + 4*x + 5, 5*x^2 + 5*x + 3]])
+            sage: rcomp = mat.basis_completion(); rcomp
+            [  2*x^2 + 1 4*x^2 + 3*x 2*x^2 + 3*x]
+            sage: mat.stack(rcomp).determinant()
+            3
+            sage: mat.basis_completion(row_wise=False)
+            []
+
+        The following matrix has rank 1 and its nonzero Smith factor is `x+3`.
+        A row-wise completion has a single nonzero row, whereas a column-wise
+        completion has two columns; in both cases, the Smith form is preserved::
+
+            sage: mat = matrix(ring, 3, 2, \
+                    [[    x^3 + x^2 + 5*x + 5,         2*x^3 + 2*x + 4], \
+                     [  3*x^3 + 2*x^2 + x + 3,   6*x^3 + 5*x^2 + x + 1], \
+                     [2*x^3 + 5*x^2 + 3*x + 4, 4*x^3 + 6*x^2 + 5*x + 6]])
+            sage: mat.smith_form(transformation=False)
+            [x + 3     0]
+            [    0     0]
+            [    0     0]
+            sage: rcomp = mat.basis_completion(); rcomp
+            [x + 1   2*x]
+            sage: ccomp = mat.basis_completion(row_wise=False); ccomp
+            [3*x + 1 4*x + 4]
+            [    2*x 5*x + 1]
+            [    6*x       x]
+            sage: rcomp.stack(mat).smith_form(transformation=False)
+            [    1     0]
+            [    0 x + 3]
+            [    0     0]
+            [    0     0]
+            sage: ccomp.augment(mat).smith_form(transformation=False)
+            [    1     0     0     0]
+            [    0     1     0     0]
+            [    0     0 x + 3     0]
+
+        Here are a few more examples, similar to the above but over fields
+        other than ``GF(7)``::
+
+            sage: ring.<x> = QQ[]
+            sage: mat = matrix([[x*(x-1)*(x-2), (x-2)*(x-3)*(x-4), (x-4)*(x-5)*(x-6)]])
+            sage: mat
+            [        x^3 - 3*x^2 + 2*x   x^3 - 9*x^2 + 26*x - 24 x^3 - 15*x^2 + 74*x - 120]
+            sage: rcomp = mat.basis_completion(algorithm="smith"); rcomp
+            [        -1/12*x - 1/12         -1/12*x + 5/12                      0]
+            [                  1/12                   1/12 1/24*x^2 - 13/24*x + 2]
+            sage: mat.stack(rcomp).determinant()
+            1
+
+            sage: mat = matrix([[x*(x-1), x*(x-2)], \
+                                [x*(x-2), x*(x-3)], \
+                                [(x-1)*(x-2), (x-1)*(x-3)]])
+            sage: mat.smith_form(transformation=False)
+            [1 0]
+            [0 x]
+            [0 0]
+            sage: ccomp = mat.basis_completion(row_wise=False, algorithm="smith")
+            sage: ccomp
+            [1/2*x - 1/2]
+            [  1/2*x - 1]
+            [1/2*x - 3/2]
+            sage: ccomp.augment(mat).smith_form(transformation=False)
+            [    1     0     0]
+            [    0     1     0]
+            [    0     0 1/2*x]
+
+            sage: ring.<y> = NumberField(x**2 - 2, "a")[]
+            sage: a = ring.base_ring().gen()
```

Comment:
```suggestion
            sage: K.<a> = NumberField(x^2 - 2)
            sage: S.<y> = K[]
```
I find this more readable.

Reply:
yes, thanks for the suggestion, I've done this (up to renaming K and S to variable names more consistent with the things around).